### PR TITLE
Wallet Action Messages

### PR DESCRIPTION
**Pull request by jdfigure**
## Description

Added functionality to support `wallet_action` method and messages. This allows wc-js and dapps to communicate information with the wallet unrelated to signing or blockchain msg broadcast. For example, this functionality is used to support a `switchToGroup` method where the dapp can ask the wallet, via `wallet_action`, to prompt the user to switch from an individual account to an individual account representing a group policy. When switching to group the wallet will send a WC update session message with the new group context, JWT in the WC account object. Then the dapp can act on behalf of the group without the user having to disconnect and reconnect their wallet.

Updated demo react-vite application to support the generic wallet action and switch group
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -78,4 +78,7 @@ module.exports = {
       },
     },
   },
+  parserOptions: {
+    project: ["tsconfig.json","examples/example-react-vite/tsconfig.json"]
+  }
```
Comment (VigM-Figure) [inline .eslintrc.js:83]: Was this needed for a specific error/issue?

Comment (jdfigure) [reply .eslintrc.js:83]: Yes, IntelliJ shows an error in every *.ts, *.tsx file when this isn't set - it was annoying. :)

File: examples/example-react-vite/src/Components/Navigation/Navigation.tsx [hunk 2]
```diff
@@ -72,6 +73,17 @@ export const Navigation: React.FC<Props> = ({ bridgeUrl }) => {
       icon: ICON_NAMES.GEAR,
       url: SEND_MESSAGE_URL,
     },
+    {
+      name: 'Send Wallet Action',
```
Comment (VigM-Figure) [inline examples/example-react-vite/src/Components/Navigation/Navigation.tsx:77]: Do we even want to expose this to dApps?  I was thinking dApps are limited to just the methods available on walletConnectService

Comment (jdfigure) [reply examples/example-react-vite/src/Components/Navigation/Navigation.tsx:77]: Yeah, I went around on that - it does provide a way for dapps to do things without waiting for specific support from us (i.e. specifically building `switchToGroup` in wc-js).

Comment (VigM-Figure) [reply examples/example-react-vite/src/Components/Navigation/Navigation.tsx:77]: Everything we add as a method should exist within walletConnectService first though, wouldn't it?  I don't think we'll ever have some hidden method that you can't call from the service.  Only benefit is not having to update the example/demo apps.

File: src/consts/windowMessages.ts [hunk 3]
```diff
@@ -3,9 +3,15 @@ export const WINDOW_MESSAGES = {
   CONNECTED: 'FWC_CONNECTED',
   // WalletConnect Disconnect
   DISCONNECT: 'FWC_DISCONNECT',
+  // WalletConnect Session Update
+  SESSION_UPDATED: 'FWC_SESSION_UPDATED',
```
Comment (VigM-Figure) [inline src/consts/windowMessages.ts:7]: This should match the method names, so this one is SWITCH_TO_GROUP

Comment (VigM-Figure) [reply src/consts/windowMessages.ts:7]: Or just remove this and instead use the below SEND_WALLET_ACTION_COMPLETE/FAILED

File: src/services/walletConnectService.ts [hunk 4]
```diff
@@ -524,6 +526,61 @@ export class WalletConnectService {
...
   };
 
+  /**
+   *
+   * @param action Wallet Action that the wallet should take (e.g. `switchToGroup`)
+   * @param payload JSON payload for wallet to handle
+   * @param description (optional) Additional information for wallet to display
+   * @param method (optional) What method is used to send this message
+   */
+  sendWalletAction = async ({
```
Comment (VigM-Figure) [inline src/services/walletConnectService.ts:536]: Thoughts on keeping this method private and instead adding the specific "actions" as service methods for dApps to use?

File: src/services/walletConnectService.ts [hunk 5]
```diff
@@ -524,6 +526,61 @@ export class WalletConnectService {
...
+  };
+
+  /**
+   *
+   * @param groupPolicyAddress the Group Policy to switch the wallet to
+   * @param description (optional) provide description to display in the wallet when
+   *        switching to group policy address
+   */
+  switchToGroup = async (groupPolicyAddress: string, description?: string) => {
+    return this.sendWalletAction({
```
Comment (VigM-Figure) [inline src/services/walletConnectService.ts:574]: Call sendWalletActionMethod here, remove this.sendWalletAction method

File: src/types/WalletConnectService.ts [hunk 6]
```diff
@@ -7,7 +7,7 @@ import type { GasPrice } from './GasPriceType';
 
 export type WalletConnectClientType = WalletConnectClient;
 
-type WCSPendingMethod = '' | 'sendMessage' | 'signJWT' | 'signHexMessage';
+type WCSPendingMethod = '' | 'sendMessage' | 'signJWT' | 'signHexMessage' | 'sendWalletAction';
```
Comment (VigM-Figure) [inline src/types/WalletConnectService.ts:10]: I think we should have a pending method named as the exact method, in this case switchToGroup

File: src/consts/windowMessages.ts (hunk 3 above)
Comment (VigM-Figure) [reply src/consts/windowMessages.ts:7]: Changed my mind again, my vote is to have the window events mirror the methods exposed to the dApp, in this case, WALLET_ACTION is not exposed, instead we should have SWITCH_TO_GROUP_COMPLETE/FAILED

File: src/types/WalletConnectService.ts [hunk 7]
```diff
@@ -7,10 +7,12 @@ import type { GasPrice } from './GasPriceType';
 
 export type WalletConnectClientType = WalletConnectClient;
 
-type WCSPendingMethod = '' | 'sendMessage' | 'signJWT' | 'signHexMessage' | 'sendWalletAction';
+type WCSPendingMethod = '' | 'sendMessage' | 'signJWT' | 'signHexMessage' | 'switchToGroup';
 
 export type WalletConnectServiceStatus = 'connected' | 'disconnected' | 'pending';
 
+export type WalletAction = 'switchToGroup';
```
Comment (VigM-Figure) [inline src/types/WalletConnectService.ts:14]: This is already a const in src/consts/walletActions.ts.
Change it to be 
```export type WalletAction = 'switchToGroup' as const;``` within walletActions.ts and you can import it as the type as well.

**Review by VigM-Figure — APPROVED** [review]

**Pull request merged**